x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/utils/ExponentialAverageCalculationContext.java

```diff
@@ -0,0 +1,182 @@
+
+    private static final TemporalUnit WINDOW_UNIT = ChronoUnit.HOURS;
+    private static final Duration WINDOW_SIZE = WINDOW_UNIT.getDuration();
+
+    private double incrementalTimeMetricValueMs;
+    private Instant latestTimestamp;
+    private Double previousExponentialAverageMs;
+
+    public ExponentialAverageCalculationContext() {
+        this(0.0, null, null);
+    }
+
+    public ExponentialAverageCalculationContext(
+            double incrementalTimeMetricValueMs,
+            @Nullable Instant latestTimestamp,
+            @Nullable Double previousExponentialAverageMs) {
+        this.incrementalTimeMetricValueMs = incrementalTimeMetricValueMs;
+        this.latestTimestamp = latestTimestamp != null ? Instant.ofEpochMilli(latestTimestamp.toEpochMilli()) : null;
+        this.previousExponentialAverageMs = previousExponentialAverageMs;
+    }
+
+    public ExponentialAverageCalculationContext(ExponentialAverageCalculationContext lhs) {
+        this(lhs.incrementalTimeMetricValueMs, lhs.latestTimestamp, lhs.previousExponentialAverageMs);
+    }
+
+    public ExponentialAverageCalculationContext(StreamInput in) throws IOException {
+        this.incrementalTimeMetricValueMs = in.readDouble();
+        this.latestTimestamp = in.readOptionalInstant();
+        this.previousExponentialAverageMs = in.readOptionalDouble();
+    }
+
+    double getIncrementalTimeMetricValueMs() {
+        return incrementalTimeMetricValueMs;
+    }
+
+    Instant getLatestTimestamp() {
+        return latestTimestamp;
+    }
+
+    Double getPreviousExponentialAverageMs() {
+        return previousExponentialAverageMs;
+    }
+
+    public Double getCurrentExponentialAverageMs() {
+        if (previousExponentialAverageMs == null || latestTimestamp == null) return incrementalTimeMetricValueMs;
+        Instant currentWindowStartTimestamp = latestTimestamp.truncatedTo(WINDOW_UNIT);
+        double alpha = Math.exp(
+            - (double) Duration.between(currentWindowStartTimestamp, latestTimestamp).toMillis() / WINDOW_SIZE.toMillis());
+        return alpha * previousExponentialAverageMs + (1 - alpha) * incrementalTimeMetricValueMs;
+    }
+
+    public void increment(double metricValueDeltaMs) {
+        incrementalTimeMetricValueMs += metricValueDeltaMs;
+    }
+
+    public void setLatestTimestamp(Instant newLatestTimestamp) {
+        Objects.requireNonNull(newLatestTimestamp);
+        if (this.latestTimestamp != null) {
+            Instant nextWindowStartTimestamp = this.latestTimestamp.truncatedTo(WINDOW_UNIT).plus(WINDOW_SIZE);
+            if (newLatestTimestamp.compareTo(nextWindowStartTimestamp) >= 0) {
+                // When we cross the boundary between windows, we update the exponential average with metric values accumulated so far in
+                // incrementalTimeMetricValueMs variable.
+                this.previousExponentialAverageMs = getCurrentExponentialAverageMs();
+                this.incrementalTimeMetricValueMs = 0.0;
```

Comment:
There is an interdependency between `setLatestTimestamp` and `increment`. It looks to me like the order in which they are called will change the result of `getCurrentExponentialAverageMs` (`increment` should be called after `setLatestTimestamp` so `incrementalTimeMetricValueMs` is non-zero??)

Is it possible to merge the 2 functions into a general update function? I would assume so as each new data point has a timestamp and metric value.

Reply:
> Is it possible to merge the 2 functions into a general update function? I would assume so as each new data point has a timestamp and metric value.

Unfortunately it is not true for the datafeed use-case.
Datafeed can execute multiple searches, each of them resulting in a data point (but without a timestamp). Then, after search results are posted to PostData API, DataCounts with latest record timestamp are returned. I couldn't find an easy way to obtain latest record timestamps right after SearchResponse is returned. That's why ExponentialAverageCalculationContext ended up having 2 methods.

Reply:
I see what you mean, I can't see a neat solution either. Can you add some comments explaining how the class should be used please.